### ISSUE TITLE
Add ability to set custom icons

### ISSUE THREAD

**Issue by dannycolin** [mozilla/multi-account-containers#2368]
Let the users add a custom icons to their containers.

**Comment by dave-kennedy** [comment]
You mentioned elsewhere that this requires work upstream. Why?

It looks like someone did this in #231. Can we not do the same thing here?

**Comment by dannycolin** [comment]
> It looks like someone did this in https://github.com/mozilla/multi-account-containers/pull/231. Can we not do the same thing here?

This icons were added manually (in the PR) in Firefox source code. This issue is to let the user uploads a custom one and the code in Firefox doesn't handle this specific use case.

**Comment by dave-kennedy** [comment]
Gotcha. So we could easily add a bunch more icons though, right?

**Comment by FlorianWendelborn** [comment]
@dave-kennedy in theory, yes. In practice, it'd be more viable to allow custom icons since it'd otherwise make firefox a competitor to icon websites in a few years 😄

**Comment by atharen** [comment]
Is having an emoji picker a reasonable substitute?

**Comment by mathstuf** [comment]
How is an emoji picker a set of hardcoded icons? Or is there some detail about the implementation that means every emoji codepoint would require a manual mapping?

**Comment by Karasiq** [comment]
> How is an emoji picker a set of hardcoded icons? Or is there some detail about the implementation that means every emoji codepoint would require a manual mapping?

I agree, that would be an ideal solution

**Comment by Jieiku** [comment]
@dannycolin Your opening post says "Let the users add a custom icons to their containers."

I am wondering if you can edit it to include more information, such as where these changes need to take place, etc.

From reading this post it seems this is simply impossible unless something upstream changes in firefox?

I am just hoping for a post that outlines the way we want the change to exist which seems to be to not hard code additional icons but allow users to add their own, which is fine. Also which project do these changes need to take place?

**Comment by dannycolin** [comment]
> From reading this post it seems this is simply impossible unless something upstream changes in firefox? I am just hoping for a post that outlines what changes need to be made and where.

You're right. It can only be implemented upstream. As I'm not a Mozilla developer, it isn't easy for me to outlines in details what's needed.

However, the tl;dr that I can give you is that you'd need to:

* Change the contextualidentity module in order to accept custom images
* Have a way to validate the image respect a set of rules (dimension, format, etc)
* Expose this feature to the webextension API
* Handle backward compatibility for addons that expect a value from the current hardcoded list of icons.

If you are considering contributing a patch, the best place to ask more question before even starting to write one line of code would be on #introduction on https://chat.mozilla.org. Make sure to mention [bug 1533542](https://bugzilla.mozilla.org/show_bug.cgi?id=1533542) and ask who you should get in touch with to mentor you.

If you never used Matrix before, read https://wiki.mozilla.org/Matrix. It has some useful information on how to join our server.

**Comment by atharen** [comment]
I'll just say, that [emoji are built-in for Firefox (ctrl+f _emoji_)](https://www.mozilla.org/en-US/firefox/50.0/releasenotes/), and [seems to have bundled font](https://searchfox.org/mozilla-release/source/browser/fonts/moz.build) as well...

Let's start from this simple premise: 

- [ ] Can the extension render a unicode/emoji symbol or is it just tofu?
  - This [extension made it work](https://addons.mozilla.org/en-US/firefox/addon/emoji-sav/?utm_source=addons.mozilla.org&utm_medium=referral&utm_content=search), from the source, looks like it [bundles some emoji fonts](https://github.com/Sav22999/emoji/tree/master/font).

From there: 

- [ ] how difficult would it be to colour emoji? or make them a single colour?
  - And if colouring is an issue, maybe don't colour the emoji character?
- [ ] The icon appears in the address bar; can unicode appear in the address bar?
- [ ] Can a custom font appear in the address bar? 
  - _Probably no, unless you stick it as a picture, presumably like the existing implementation..._

I dunno, overall seems like something that should go through without issue; mainly doubtful of the latter point, and even then, one can use the system fonts (_or maybe twemoji?_), which might not be consistent across OSs, I don't think that is going to affect a lot of users.

**Comment by atharen** [comment]
For the record, the only reason I don't like custom icons, is because they wouldn't be portable - I don't want to load my multi-account container icons on each fresh Firefox install.

And if there is some free service for extension backups, I doubt the provider intends it to be for bitmaps, probably just for configs.

**Comment by MikeDacre** [comment]
@dannycolin This issue has existed for a long time and seems somewhat stalled by upstream.

Would you be amenable to the addition of a few more SVG icons to the current set (maybe 6 more)? 

I would be happy to add them myself in a PR.

Also open to suggestions on which icons to make and how many.

**Comment by mathstuf** [comment]
Selfishly, icons that would be nice to replace the circles I am using:

- newspaper/magnifying glass/magazine for "news"
- something graph-like for "social"
- wrench for "tools" or "code"
- controller for "game"
- sports thing (ball, bike, whatever) for "fitness"
- something for temporary tabs (smoke, cloud, clock?)

Just suggestions; take or leave as anyone sees fit (or according to difficulty).

**Comment by dannycolin** [comment]
@MikeDacre it's been triaged as P5 (See [bug 1533542](https://bugzilla.mozilla.org/show_bug.cgi?id=1533542)) meaning Firefox developers (upstream) are open to accept a patch from a contributor. 

If you wish to do so, you can read the documentation on https://firefox-source-docs.mozilla.org/contributing/contribution_quickref.html#firefox-contributors-quick-reference. This will help you get a working development environment to build Firefox locally and work on your patch.

If at any point you're stuck, please don’t hesitate to ask at https://chat.mozilla.org/ in the [#introduction](https://chat.mozilla.org/#/room/#introduction:mozilla.org) channel for any questions on the tools used to work on Firefox codebase. You can also join [#containers](https://chat.mozilla.org/#/room/#containers:mozilla.org) if you want to discuss further about this patch.

**Comment by MikeDacre** [comment]
@dannycolin Thanks!

I am not sure if I'll have time to do all of that though, I was actually thinking of replicating PR #231 and adding a small number of new hard coded icons. Would that be acceptable for now?

I will see if I have time to work on the actual new feature folks are asking for the way you suggest also.

**Comment by dannycolin** [comment]
> I was actually thinking of replicating PR #231 and adding a small number of new hard coded icons. Would that be acceptable for now?

We can't implement it like in #231 because this code is now part of Firefox itself.

This is where the CSS code goes and svgs are placed in `/content` directory:
https://searchfox.org/mozilla-central/source/browser/components/contextualidentity/content/usercontext.css#50

You'll need to update this two lists (only adding the new icon names):
* https://searchfox.org/mozilla-central/source/toolkit/components/extensions/parent/ext-contextualIdentities.js#44
* https://searchfox.org/mozilla-central/source/browser/components/preferences/dialogs/containers.js#26

Then, you'll need to add the icons where needed in Multi-Account Containers. It's a "good-first-bug" in term of complexity but it indeed needs time to update everything.

Another thing is the icons you're going to use would probably need to be vetted by someone at Mozilla mostly to make sure the license allow them to distribute the icons.

**Comment by MikeDacre** [comment]
FYI, work has become sufficiently intense that I will be unlikely to have time to do this in the immediate foreseeable future. If anyone else has time, go ahead. Otherwise, @dannycolin, fine to close as "won't fix" if you don't want to clutter the open issues. If I have time later I'll circle back and just create a pull request directly.

**Comment by Fred-Vatin** [comment]
It would be nice if it uses an icon picker filtering through a built-in font-awesome set for instance.

**Comment by sinasalek** [comment]
I was expecting to be able to upload my own icon for containers but strangely enough that's no possible! And there isn't any non messy way of making it possible. So I'm very much interested for this to happen.
Probably the most flexible option is to let extensions customize these icons

**Comment by babs** [comment]
It would be nice to be able to customize colors too :)

**Comment by HomeHeartTherapy** [comment]
I REALLY want to be able to easily import my own icons per container.  For example, I want to use the REAL app logos for these containers:  1. Python 2. Obsidian 3. Docker and so on and so forth.

Also, it would be great to have a shortcut in the extension that would allow us to do custom "screen capture" of an area of the screen that we select.  Therefore, if we work for....let's say McDonald's and we have a container dedicated to McDonald's.  We could go to their homepage, hit the shortcut key, make the screen selection JUST big enough to get a high rez "screenshot" of the Company logo, then the extension makes it the right size and shape (I'm assuming a circle), and then we can assign it to that container.  We could repeat in a different container if we had a second place of employment...say...NASA!  (that's right!  I'm gonna be eating french fries...IN SPACE!!!).  Wouldn't that make for a beautiful and proper container logo?

That way every container would have a proper, high rez logo for it whether it is a "subject container" or a "work container" etc etc. 

Isn't this a universally wanted and requested feature?  I've JUST started learning how to code so forgive me if this is a stupid or unreasonable request.  I'm just asking as a end user that really loves this extension but want it to look polished.

Thoughts?

By the way...I just want to say thank you for everything that you have already done.  Even as is...this extension is wonderful and we VERY MUCH appreciate it!

**Comment by rublev** [comment]
omfg please custom colors and icons is that rly too hard. sidebery does it perfectly

<img width="451" height="305" alt="Image" src="https://github.com/user-attachments/assets/61681298-6fd7-48c3-9a58-828cfbc0f144" />
